subgraphs/portal/src/mappings/handlers.ts

```diff
@@ -0,0 +1,478 @@
+  truncateAddress,
+} from "./helpers";
+import { getUsdPrice, getUsdPricePerToken } from "../prices";
+
+import { SDK } from "../sdk/protocols/bridge";
+import { TokenInitializer, TokenParams } from "../sdk/protocols/bridge/tokens";
+import { BridgeConfig } from "../sdk/protocols/bridge/config";
+import {
+  BridgePermissionType,
+  BridgePoolType,
+  CrosschainTokenType,
+} from "../sdk/protocols/bridge/enums";
+import { TokenPricer } from "../sdk/protocols/config";
+import { bigIntToBigDecimal } from "../sdk/util/numbers";
+import { BIGINT_ZERO, ZERO_ADDRESS } from "../sdk/util/constants";
+
+import {
+  Bridge,
+  CompleteTransferAndUnwrapETHCall,
+  CompleteTransferAndUnwrapETHWithPayloadCall,
+  CompleteTransferCall,
+  CompleteTransferWithPayloadCall,
+} from "../../generated/Core/Bridge";
+import { Core, LogMessagePublished } from "../../generated/Core/Core";
+import { _ERC20 } from "../../generated/Core/_ERC20";
+import { Token } from "../../generated/schema";
+
+const conf = new BridgeConfig(
+  NetworkConfigs.getFactoryAddress().toHexString(),
+  NetworkConfigs.getProtocolName(),
+  NetworkConfigs.getProtocolSlug(),
+  BridgePermissionType.PRIVATE,
+  Versions
+);
+
+class Pricer implements TokenPricer {
+  getTokenPrice(token: Token): BigDecimal {
+    const pricedToken = Address.fromBytes(token.id);
+
+    return getUsdPricePerToken(pricedToken).usdPrice;
+  }
+
+  getAmountValueUSD(token: Token, amount: BigInt): BigDecimal {
+    const pricedToken = Address.fromBytes(token.id);
+    const _amount = bigIntToBigDecimal(amount, token.decimals);
+
+    return getUsdPrice(pricedToken, _amount);
+  }
+}
+
+class TokenInit implements TokenInitializer {
+  getTokenParams(address: Address): TokenParams {
+    const erc20 = _ERC20.bind(address);
+    const name = erc20.name();
+    const symbol = erc20.symbol();
+    const decimals = erc20.decimals().toI32();
+
+    return {
+      name,
+      symbol,
+      decimals,
+    };
+  }
+}
+
+export function handleSwapOut(event: LogMessagePublished): void {
+  const payload = event.params.payload;
+
+  let amount: BigInt;
+  let tokenAddress: Bytes;
+  let tokenChain: i32;
+  let to: Bytes;
+  let toChain: i32;
+
+  const bridgeContract = Bridge.bind(NetworkConfigs.getBridgeAddress());
+  const parseTransferCall = bridgeContract.try_parseTransfer(payload);
+  if (parseTransferCall.reverted) {
+    const parseTransferWithPayloadCall =
+      bridgeContract.try_parseTransferWithPayload(payload);
+    if (parseTransferWithPayloadCall.reverted) {
+      log.warning("[handleSwapOut] tx_hash: {} failed to parse payload: {}", [
+        event.transaction.hash.toHexString(),
+        payload.toHexString(),
+      ]);
+      return;
+    }
+
+    const transferStruct = parseTransferWithPayloadCall.value;
+
+    amount = transferStruct.amount;
+    tokenAddress = transferStruct.tokenAddress;
+    tokenChain = transferStruct.tokenChain;
+    to = transferStruct.to;
+    toChain = transferStruct.toChain;
```

Comment:
moving these outside the block would require me to define `transferStruct` outside the block too, but that's difficult because `transferStruct` could be of either `Bridge__parseTransferWithPayloadResultTransferStruct` or `Bridge__parseTransferResultTransferStruct` types, and assembly script doesn't allow me to define with union types

Reply:
Ah I understand